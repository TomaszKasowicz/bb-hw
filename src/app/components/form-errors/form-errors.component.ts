import { Component, Input, OnInit } from '@angular/core';
import { AbstractControl } from '@angular/forms';

export interface ErrorMessages {
  [key: string]: string;
}

@Component({
  selector: 'app-form-errors',
  template: `
    <ng-container *ngIf="ctrl?.dirty && ctrl?.errors">
      <div class="error" *ngFor="let error of ctrl?.errors | keyvalue">
        {{ errorMessages[error.key] || error.key }}
      </div>
    </ng-container>
  `,
  styles: [
    `
      :host {
        width: 100%;
      }

      .error {
        font-size: 0.75em;
        width: 100%;
        color: var(--color-peachtree-orange);
      }
    `
  ]
})
export class FormErrorsComponent {
  @Input()
  public ctrl: AbstractControl | null;

  @Input()
  public errorMessages: ErrorMessages = {};
}
