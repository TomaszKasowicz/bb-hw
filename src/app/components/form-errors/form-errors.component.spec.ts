import { createComponentFactory } from '@ngneat/spectator/jest';
import { byText, Spectator } from '@ngneat/spectator';
import { FormErrorsComponent } from './form-errors.component';
import { FormControl } from '@angular/forms';
describe('FormErrorsComponent', () => {
  let spectator: Spectator<FormErrorsComponent>;
  const createCompoent = createComponentFactory(FormErrorsComponent);

  beforeEach(() => {
    spectator = createCompoent();
  });

  it('Should create', () => {
    expect(spectator.component).toBeTruthy();
  });

  it('Should not render any errors if control is pristine', () => {
    const ctrl = new FormControl('');
    ctrl.setErrors({ someError: true }, { emitEvent: false });
    const errorMessages = { someError: 'Some Error Message' };
    spectator.setInput({ ctrl, errorMessages });

    spectator.detectChanges();
    expect(spectator.fixture).toMatchSnapshot();

    const divElements = spectator.queryAll('div');
    expect(divElements.length).toEqual(0);
  });

  it('Should not render any errors if control is dirty but valid', () => {
    const ctrl = new FormControl('');
    ctrl.markAsDirty();
    const errorMessages = { someError: 'Some Error Message' };

    spectator.setInput({ ctrl, errorMessages });
    spectator.detectChanges();

    expect(spectator.fixture).toMatchSnapshot();

    const divElements = spectator.queryAll('div');
    expect(divElements.length).toEqual(0);
  });

  it('Should render message when ctrl has errors and error Message is available', () => {
    const ctrl = new FormControl('');
    ctrl.setErrors({ someError: true }, { emitEvent: false });
    ctrl.markAsDirty();

    const errorMessages = { someError: 'Some Error Message' };

    spectator.setInput({ ctrl, errorMessages });
    spectator.detectChanges();
    expect(spectator.fixture).toMatchSnapshot();
    expect(spectator.query(byText('Some Error Message'))).toBeTruthy();
  });

  it('Should render error key when error message is not available', () => {
    const ctrl = new FormControl('');
    ctrl.setErrors({ someError: true }, { emitEvent: false });
    ctrl.markAsDirty();

    const errorMessages = { someOtherError: 'Some Error Message' };

    spectator.setInput({ ctrl, errorMessages });
    spectator.detectChanges();
    expect(spectator.fixture).toMatchSnapshot();
    expect(spectator.query(byText('someError'))).toBeTruthy();
  });
});
