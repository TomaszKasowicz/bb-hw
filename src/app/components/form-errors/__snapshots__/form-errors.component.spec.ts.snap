// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormErrorsComponent Should not render any errors if control is dirty but valid 1`] = `
<app-form-errors
  __ngContext__={[Function LRootView]}
  ctrl={[Function FormControl]}
  errorMessages={[Function Object]}
>
  
</app-form-errors>
`;

exports[`FormErrorsComponent Should not render any errors if control is pristine 1`] = `
<app-form-errors
  __ngContext__={[Function LRootView]}
  ctrl={[Function FormControl]}
  errorMessages={[Function Object]}
>
  
</app-form-errors>
`;

exports[`FormErrorsComponent Should render error key when error message is not available 1`] = `
<app-form-errors
  __ngContext__={[Function LRootView]}
  ctrl={[Function FormControl]}
  errorMessages={[Function Object]}
>
  <div
    class="error"
  >
     someError 
  </div>
</app-form-errors>
`;

exports[`FormErrorsComponent Should render message when ctrl has errors and error Message is available 1`] = `
<app-form-errors
  __ngContext__={[Function LRootView]}
  ctrl={[Function FormControl]}
  errorMessages={[Function Object]}
>
  <div
    class="error"
  >
     Some Error Message 
  </div>
</app-form-errors>
`;
